File: backend/modules/task_management/domain/entity/task.go
package domain

import (
	"errors"
	securityEntity "stm/modules/security/domain/entity"
)

type TaskStatus struct {
	Id    uint   `gorm:"primaryKey;autoIncrement" json:"id"`
	Name  string `gorm:"size:100" json:"name"`
	Color string `gorm:"size:7" json:"color"`
}

func (TaskStatus) TableName() string {
	return "task_status"
}

type Task struct {
	Id           uint                `gorm:"primaryKey; autoIncrement" json:"id"`
	Title        string              `gorm:"size:100" json:"title"`
	Description  string              `gorm:"size:255" json:"description"`
	Status       TaskStatus          `json:"status"`
	StatusId     uint                `json:"statusId"`
	CreatedBy    securityEntity.User `json:"-"`
	CreatedById  string              `json:"createdBy"`
	AssignedTo   securityEntity.User `json:"-"`
	AssignedToId string              `gorm:"default:null" json:"assignedTo"`
}

func (u *Task) Validate() error {
	if u.Title == "" {
		return errors.New("must specify task title")
	}

	if u.Description == "" {
		return errors.New("must specify task description")
	}

	if u.StatusId == 0 && u.Status.Id == 0 {
		return errors.New("must specify status")
	}

	if u.Id == 0 && u.CreatedById == "" && u.CreatedBy.Id == "" {
		return errors.New("must specify created by")
	}

	return nil
}

File: frontend/src/modules/task_management/domain/entity/task.ts
import { validator } from "@/modules/shared/domain/errors";

export type Task  = {
    id: number;
    title: string;
    description: string;
    statusId: number;
    assignedTo: string;   
}

export type TaskRecord  = {
    id: number;
    title: string;
    description: string;
    statusId: number;   
    status: string;
    color: string;
    createdBy: string; 
    assignedTo: string;     
}

export type TaskStatus = {
    id: number;
    name: string;
    color: string;
}

export type TaskFilter = CriteriaFilter & {
    search?: string;
    statusId: number; 
    createdBy: string; 
    assignedTo: string; 
}

export const validateTask = (task: Task) => {
    const dataValidation = [
        { validation: !task.title, error: 'TASK.must_specify_title', tag: 'title' },
        { validation: !task.description, error: 'TASK.must_specify_description', tag: 'description' },
        { validation: !task.statusId, error: 'TASK.must_specify_status', tag: 'status' },
    ];

    validator(dataValidation);
}
